### PR TITLE
[WIP] Add recover from diagnostic errors

### PR DESCRIPTION
**Pull request by Deleh**
Adds recovery from diagnostic errors to `halt_detector`.

### CODE REVIEW

**Review by fmessmer — COMMENTED** [review]
not sure this is generic enough - although `halt_detector` is currently only used for kevins
maybe it needs more configurability so that the diagnostics recovery and/or which error codes to react to can be configured specifically to the robot

**Comment by Deleh** [comment]
> not sure this is generic enough - although `halt_detector` is currently only used for kevins
> maybe it needs more configurability so that the diagnostics recovery and/or which error codes to react to can be configured specifically to the robot

This should be given by setting the _diagnostic recovery trigger_ parameter per robot. The launch file needs to be adjusted to not overwrite it

File: cob_base_controller_utils/launch/halt_detector.launch [hunk 1]
```diff
@@ -7,6 +7,7 @@
   <arg name="component_name" default="base" />
 
   <group ns="$(arg component_name)">
+    <rosparam param="halt_detector/diagnostic_recovery_trigger">['ff20#0']</rosparam>
```
Comment (floweisshardt) [inline cob_base_controller_utils/launch/halt_detector.launch]: make this an entry (list of strings) in a robot specific yaml file in mojin_hardware_config

**Review by floweisshardt — CHANGES_REQUESTED** [review]

File: cob_base_controller_utils/config/halt_detector.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+diagnostic_recovery_trigger: []
```
Comment (floweisshardt) [inline cob_base_controller_utils/config/halt_detector.yaml]: this yaml file lives in mojin_hardware_config. remove here

File: cob_base_controller_utils/launch/halt_detector.launch [hunk 3]
```diff
@@ -1,18 +1,8 @@
 <launch>
-  <arg name="stop_timeout" default="10.0" />
-  <arg name="stop_threshold" default="0.001" />
-  <arg name="stuck_timeout" default="1.0" />
-  <arg name="stuck_threshold" default="1.57" />
-  <arg name="recover_after_stuck_timeout" default="3.0" />
+  <arg name="config_file" default="$(find cob_base_controller_utils)/config/halt_detector.yaml" />
```
Comment (floweisshardt) [inline cob_base_controller_utils/launch/halt_detector.launch]: remove default

File: cob_base_controller_utils/launch/halt_detector.launch [hunk 4]
```diff
@@ -1,18 +1,8 @@
...
-
   <group ns="$(arg component_name)">
-    <node pkg="cob_base_controller_utils" type="cob_halt_detector" name="halt_detector" output="screen">
-        <param name="stop_timeout" type="double" value="$(arg stop_timeout)"/>
-        <param name="stop_threshold" type="double" value="$(arg stop_threshold)"/>
-        <param name="stuck_timeout" type="double" value="$(arg stuck_timeout)"/>
-        <param name="stuck_threshold" type="double" value="$(arg stuck_threshold)"/>
-        <param name="recover_after_stuck_timeout" type="double" value="$(arg recover_after_stuck_timeout)"/>
-    </node>
+    <rosparam command="load" ns="halt_detector" file="$(arg config_file)" />
```
Comment (floweisshardt) [inline cob_base_controller_utils/launch/halt_detector.launch]: move rosparam inside node tag

**Review by floweisshardt — CHANGES_REQUESTED** [review]

**Review by floweisshardt — APPROVED** [review]

**Pull request merged**